File: forms/first-responder-assessment/back-end-with-loco/src/engine/responder_grader.rs
//! Responder grader module.

use super::flagged_issues::detect_additional_flags;
use super::responder_rules::all_rules;
use super::types::{
    AssessmentData, CompetencyLevel, DomainLevels, FiredRule, FitnessDecision, GradingResult,
    RiskLevel,
};
use super::utils::aggregate_competency;

/// Pure function: evaluates all first responder rules against assessment
/// data. Returns domain competency levels, overall fitness decision, and all
/// fired rules. Mirrors the front-end `calculateResponderGrade` function.
pub fn calculate_responder_grade(data: &AssessmentData) -> GradingResult {
    let fired_rules = run_rules(data);
    let domain_levels = derive_domain_levels(data);
    let overall_competency = derive_overall_competency(&domain_levels);
    let overall_fitness = derive_overall_fitness(data, &fired_rules, &overall_competency);
    let overall_risk = derive_overall_risk(&fired_rules, &overall_competency);
    let additional_flags = detect_additional_flags(data);

    GradingResult {
        overall_competency,
        overall_fitness,
        overall_risk,
        domain_levels,
        fired_rules,
        additional_flags,
        timestamp: chrono::Utc::now().to_rfc3339(),
    }
}

/// Evaluate every rule against the data, returning the rules that fired.
pub fn run_rules(data: &AssessmentData) -> Vec<FiredRule> {
    let mut out = Vec::new();
    for rule in all_rules() {
        if (rule.evaluate)(data) {
            out.push(FiredRule {
                id: rule.id.to_string(),
                domain: rule.domain.to_string(),
                description: rule.description.to_string(),
                grade: rule.grade,
            });
        }
    }
    out
}

/// Derive domain competency levels from raw assessment data.
pub fn derive_domain_levels(d: &AssessmentData) -> DomainLevels {
    DomainLevels {
        physical_fitness: aggregate_competency(&[
            d.physical_fitness.cardiovascular_fitness.as_str(),
            d.physical_fitness.muscular_strength.as_str(),
            d.physical_fitness.manual_handling_competency.as_str(),
            d.physical_fitness.flexibility_mobility.as_str(),
            d.physical_fitness.balance_coordination.as_str(),
        ]),
        clinical_skills: aggregate_competency(&[
            d.clinical_skills.basic_life_support.as_str(),
            d.clinical_skills.advanced_life_support.as_str(),
            d.clinical_skills.airway_management.as_str(),
            d.clinical_skills.patient_assessment.as_str(),
            d.clinical_skills.trauma_assessment.as_str(),
            d.clinical_skills.triage_competency.as_str(),
            d.clinical_skills.drug_administration.as_str(),
        ]),
        equipment_vehicle: aggregate_competency(&[
            d.equipment_vehicle.defibrillator_competency.as_str(),
            d.equipment_vehicle.monitor_competency.as_str(),
            d.equipment_vehicle.stretcher_competency.as_str(),
            d.equipment_vehicle.ambulance_driving.as_str(),
            d.equipment_vehicle.equipment_check_competency.as_str(),
        ]),
        communication: aggregate_competency(&[
            d.communication_skills.patient_communication.as_str(),
            d.communication_skills.handover_competency.as_str(),
            d.communication_skills.documentation_competency.as_str(),
            d.communication_skills.safeguarding_awareness.as_str(),
        ]),
        psychological: aggregate_competency(&[
            d.psychological_readiness.stress_management.as_str(),
            d.psychological_readiness
                .decision_making_under_pressure
                .as_str(),
            d.psychological_readiness.emotional_regulation.as_str(),
        ]),
    }
}

/// Derive overall competency from the worst domain level.
fn derive_overall_competency(levels: &DomainLevels) -> CompetencyLevel {
    aggregate_competency(&[
        levels.physical_fitness.as_str(),
        levels.clinical_skills.as_str(),
        levels.equipment_vehicle.as_str(),
        levels.communication.as_str(),
        levels.psychological.as_str(),
    ])
}

/// Derive overall fitness decision. If the assessor has already set a
/// decision in `fitness_decision.overallFitness` use it directly; otherwise
/// auto-derive from the fired rules and overall competency.
fn derive_overall_fitness(
    data: &AssessmentData,
    fired_rules: &[FiredRule],
    overall_competency: &str,
) -> FitnessDecision {
    if !data.fitness_decision.overall_fitness.is_empty() {
        return data.fitness_decision.overall_fitness.clone();
    }

    let max_grade = fired_rules.iter().map(|r| r.grade).max().unwrap_or(0);

    if max_grade >= 4 {
        return "permanently-unfit".to_string();
    }
    if max_grade >= 3 || overall_competency == "not-competent" {
        return "temporarily-unfit".to_string();
    }
    if max_grade >= 2 || overall_competency == "developing" {
        return "fit-with-restrictions".to_string();
    }
    "fit-for-duty".to_string()
}

/// Derive overall risk from fired rules and overall competency.
fn derive_overall_risk(fired_rules: &[FiredRule], overall_competency: &str) -> RiskLevel {
    let max_grade = fired_rules.iter().map(|r| r.grade).max().unwrap_or(0);

    if max_grade >= 4 || overall_competency == "not-competent" {
        return "critical".to_string();
    }
    if max_grade >= 3 {
        return "high".to_string();
    }
    if max_grade >= 2 || overall_competency == "developing" {
        return "moderate".to_string();
    }
    "low".to_string()
}

File: forms/first-responder-assessment/front-end-with-html/js/responder-grader.js
// First Responder grader. Pure functions: take an `AssessmentData` object
// and produce a `GradingResult` (overall competency, fitness decision,
// risk level, domain breakdown, fired rules, additional flags, timestamp).
//
// Ported directly from `responder-grader.ts`.

/**
 * @typedef {import('./types.js').AssessmentData} AssessmentData
 * @typedef {import('./types.js').CompetencyLevel} CompetencyLevel
 * @typedef {import('./types.js').FitnessDecision} FitnessDecision
 * @typedef {import('./types.js').RiskLevel} RiskLevel
 * @typedef {import('./types.js').FiredRule} FiredRule
 * @typedef {import('./types.js').GradingResult} GradingResult
 */

// Wrapped in an IIFE; published via window.FirstResponderAssessment.
(function () {
'use strict';
window.FirstResponderAssessment = window.FirstResponderAssessment || {};

const NS = window.FirstResponderAssessment;
const { responderRules, aggregateCompetency } = NS;

/**
 * Pure function: evaluates all first responder rules against assessment data.
 * Returns domain competency levels, overall fitness decision, and all fired rules.
 *
 * @param {AssessmentData} data
 * @returns {GradingResult}
 */
function calculateResponderGrade(data) {
  /** @type {FiredRule[]} */
  const firedRules = [];

  for (const rule of responderRules) {
    try {
      if (rule.evaluate(data)) {
        firedRules.push({
          id: rule.id,
          domain: rule.domain,
          description: rule.description,
          grade: rule.grade
        });
      }
    } catch (e) {
      // Rule evaluation failed - log for debugging but continue grading
      console.warn(`Responder rule ${rule.id} evaluation failed:`, e);
    }
  }

  // Determine domain competency levels
  const domainLevels = deriveDomainLevels(data);

  // Determine overall competency (worst domain)
  const overallCompetency = deriveOverallCompetency(domainLevels);

  // Determine overall fitness decision
  const overallFitness = deriveOverallFitness(data, firedRules, overallCompetency);

  // Determine overall risk from worst fired rule grade
  const overallRisk = deriveOverallRisk(firedRules, overallCompetency);

  const additionalFlags = NS.detectAdditionalFlags(data);

  return {
    overallCompetency,
    overallFitness,
    overallRisk,
    domainLevels,
    firedRules,
    additionalFlags,
    timestamp: new Date().toISOString()
  };
}

/** Derive domain competency levels from assessment data. */
function deriveDomainLevels(data) {
  return {
    physicalFitness: aggregateCompetency([
      data.physicalFitness.cardiovascularFitness,
      data.physicalFitness.muscularStrength,
      data.physicalFitness.manualHandlingCompetency,
      data.physicalFitness.flexibilityMobility,
      data.physicalFitness.balanceCoordination
    ]),
    clinicalSkills: aggregateCompetency([
      data.clinicalSkills.basicLifeSupport,
      data.clinicalSkills.advancedLifeSupport,
      data.clinicalSkills.airwayManagement,
      data.clinicalSkills.patientAssessment,
      data.clinicalSkills.traumaAssessment,
      data.clinicalSkills.triageCompetency,
      data.clinicalSkills.drugAdministration
    ]),
    equipmentVehicle: aggregateCompetency([
      data.equipmentVehicle.defibrillatorCompetency,
      data.equipmentVehicle.monitorCompetency,
      data.equipmentVehicle.stretcherCompetency,
      data.equipmentVehicle.ambulanceDriving,
      data.equipmentVehicle.equipmentCheckCompetency
    ]),
    communication: aggregateCompetency([
      data.communicationSkills.patientCommunication,
      data.communicationSkills.handoverCompetency,
      data.communicationSkills.documentationCompetency,
      data.communicationSkills.safeguardingAwareness
    ]),
    psychological: aggregateCompetency([
      data.psychologicalReadiness.stressManagement,
      data.psychologicalReadiness.decisionMakingUnderPressure,
      data.psychologicalReadiness.emotionalRegulation
    ])
  };
}

/** Derive overall competency from the worst domain level. */
function deriveOverallCompetency(domainLevels) {
  const levels = Object.values(domainLevels).filter((l) => l !== '');
  if (levels.length === 0) return '';
  return aggregateCompetency(levels);
}

/** Derive overall fitness decision. */
function deriveOverallFitness(data, firedRules, overallCompetency) {
  // If assessor has already set a decision, use it
  if (data.fitnessDecision.overallFitness !== '') {
    return data.fitnessDecision.overallFitness;
  }

  // Auto-derive based on rules and competency
  const maxGrade = firedRules.length > 0 ? Math.max.apply(null, firedRules.map((r) => r.grade)) : 0;

  // Any grade 4 rule = permanently or temporarily unfit
  if (maxGrade >= 4) return 'permanently-unfit';

  // Any grade 3 rule or not-competent = temporarily unfit
  if (maxGrade >= 3 || overallCompetency === 'not-competent') return 'temporarily-unfit';

  // Any grade 2 rule or developing = fit with restrictions
  if (maxGrade >= 2 || overallCompetency === 'developing') return 'fit-with-restrictions';

  // Otherwise = fit for duty
  return 'fit-for-duty';
}

/** Derive overall risk from fired rules and competency. */
function deriveOverallRisk(firedRules, overallCompetency) {
  const maxGrade = firedRules.length > 0 ? Math.max.apply(null, firedRules.map((r) => r.grade)) : 0;

  if (maxGrade >= 4 || overallCompetency === 'not-competent') return 'critical';
  if (maxGrade >= 3) return 'high';
  if (maxGrade >= 2 || overallCompetency === 'developing') return 'moderate';
  return 'low';
}

Object.assign(window.FirstResponderAssessment, {
  calculateResponderGrade
});
})();
